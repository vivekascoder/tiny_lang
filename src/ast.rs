// Types for our language.

use std::{fmt::Display, rc::Rc};

pub type Program = Vec<Statement>;
pub type BlockStatement = Vec<Statement>;

// Precedence order
#[derive(PartialEq, PartialOrd, Debug, Clone)]
pub enum Precedence {
    Lowest,
    Equals,      // ==
    LessGreater, // > or <
    Sum,         // +
    Product,     // *
    Prefix,      // -X or !X
    Call,        // myFunction(x)
    Index,       // array[index]
    LRShift,
    BitwiseAnd,
    BitwiseXor,
    BitwiseOr,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Expr {
    Ident(Ident),
    Infix(Infix, Box<Expr>, Box<Expr>),
    Prefix(Prefix, Box<Expr>),
    Literal(Literal),
    Call(FunctionCall),
}

#[derive(Clone, Debug, PartialEq)]
pub struct FunctionCall {
    pub parameters: Vec<Expr>,
    pub name: Rc<str>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct NativeFunctionCall {
    pub parameters: Vec<ExprResult>,
    pub name: Rc<str>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum ExprResult {
    Bool(bool),
    UnsignedInteger(usize),
    SignedInteger(isize),
    Char(char),
    Return(Box<ExprResult>),
    Void,
}

impl Display for ExprResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Bool(b) => write!(f, "{}", b),
            Self::UnsignedInteger(i) => write!(f, "{}", i),
            Self::SignedInteger(i) => write!(f, "{}", i),
            Self::Void => write!(f, ""),
            Self::Return(v) => write!(f, "{}", *v),
            Self::Char(c) => write!(f, "{}", c),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum MemoryObject {
    ExprResult(ExprResult),
    Function(Function),
}

#[derive(Clone, Debug, PartialEq)]
pub enum Infix {
    Plus,
    Minus,
    Divide,
    Multiply,
    DoubleEqual,
    NotEqual,
    GreaterThanEqual,
    GreaterThan,
    LessThanEqual,
    LessThan,

    // Bitwise operations
    LeftShift,
    RightShift,
    BitwiseAnd,
    BitwiseXor,
    BitwiseOr,
}

#[derive(PartialEq, Clone, Debug)]
pub enum Prefix {
    Plus,
    Minus,
    Not,
}

#[derive(PartialEq, Clone, Debug)]
pub enum Literal {
    UnsignedInteger(usize),
    Bool(bool),
    Char(char),
    String(Rc<str>),
}

// enum TVec {
//     LiteralVec<Literal>
// }

// WASM specific choices
// pub enum Literal {
//     I32(i32),
//     I64(i64),
//     F32(i32),
//     F64(i64),
//     Bool(bool),
//     Char(char),
// }

#[derive(Clone, Debug, PartialEq)]
pub struct Ident(pub Rc<str>);

#[derive(Clone, Debug, PartialEq)]
pub enum Statement {
    Let(Ident, Option<Type>, Expr),
    Mutate(Ident, Expr),
    Function(Function),
    Return(Expr),
    If(Condition),
    Expr(Expr),
    While(While),
}

#[derive(Clone, Debug, PartialEq)]
pub struct While {
    pub condition: Expr,
    pub body: BlockStatement,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Condition {
    pub condition: Expr,
    pub if_body: BlockStatement,
    pub else_body: Option<BlockStatement>,
}

#[derive(Clone, Debug, PartialEq, Copy)]
pub enum Type {
    UnsignedInteger,
    SignedInteger,
    Bool,
    Char,
    String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Function {
    pub name: Rc<str>,
    pub params: Vec<(Ident, Type)>,
    pub return_type: Option<Type>,
    pub body: Vec<Statement>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    EOF,
    Identifier(Rc<str>),
    Usize(usize),
    Boolean(bool),
    String(Rc<str>),
    Colon,
    SemiColon,

    // Operators
    Equal,
    DoubleEqual,
    Plus,
    Minus,
    Multiply,
    Divide,
    Mod,
    NotEqual,
    LessThan,
    LessThanEqual,
    GreaterThan,
    GreaterThanEqual,
    Bang,
    LParen,
    RParen,
    LBrace, // {
    RBrace, // }
    Comma,
    SQuote,     // '
    DQuote,     // "
    BSlash,     // \
    NewLine,    // \n
    Tab,        // \t
    Space,      // ' '
    LeftShift,  // <<
    RightShift, // >>
    Ampersand,  // &
    Carrot,     // ^
    Pipe,       // |

    // Keywords
    KeywordLet,
    KeywordUsize,
    KeywordIsize,
    KeywordBool,
    KeywordFun,
    KeywordReturn,
    KeywordIf,
    KeywordElse,
    KeywordVoid,
    KeywordChar,
    KeywordWhile,
    KeywordStr,

    // Symbol
    SymbolReturn,
}

/// implement the Debug trait for all the variants of TokenType
impl Display for TokenType {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            TokenType::EOF => write!(f, "EOF"),
            TokenType::Identifier(s) => write!(f, "{}", s),
            TokenType::Usize(i) => write!(f, "{}", i),
            TokenType::Boolean(b) => write!(f, "{}", b),
            TokenType::Colon => write!(f, ":"),
            TokenType::SemiColon => write!(f, ";"),
            TokenType::Equal => write!(f, "="),
            TokenType::DoubleEqual => write!(f, "=="),
            TokenType::Plus => write!(f, "+"),
            TokenType::Minus => write!(f, "-"),
            TokenType::Multiply => write!(f, "*"),
            TokenType::Divide => write!(f, "/"),
            TokenType::Mod => write!(f, "%"),
            TokenType::NotEqual => write!(f, "!="),
            TokenType::LessThan => write!(f, "<"),
            TokenType::LessThanEqual => write!(f, "<="),
            TokenType::GreaterThan => write!(f, ">"),
            TokenType::GreaterThanEqual => write!(f, ">="),
            TokenType::Bang => write!(f, "!"),
            TokenType::LParen => write!(f, "("),
            TokenType::RParen => write!(f, ")"),
            TokenType::LBrace => write!(f, "{{"),
            TokenType::RBrace => write!(f, "}}"),
            TokenType::Comma => write!(f, ","),
            TokenType::SQuote => write!(f, "'"),
            TokenType::DQuote => write!(f, "\""),
            TokenType::BSlash => write!(f, "\\"),
            TokenType::NewLine => write!(f, "\n"),
            TokenType::Tab => write!(f, "\t"),
            TokenType::Space => write!(f, " "),
            TokenType::KeywordLet => write!(f, "let"),
            TokenType::KeywordUsize => write!(f, "usize"),
            TokenType::KeywordIsize => write!(f, "isize"),
            TokenType::KeywordBool => write!(f, "bool"),
            TokenType::KeywordFun => write!(f, "fun"),
            TokenType::KeywordReturn => write!(f, "return"),
            TokenType::KeywordIf => write!(f, "if"),
            TokenType::KeywordElse => write!(f, "else"),
            TokenType::KeywordVoid => write!(f, "void"),
            TokenType::KeywordChar => write!(f, "chat"),
            TokenType::KeywordWhile => write!(f, "while"),
            TokenType::SymbolReturn => write!(f, "=>"),
            TokenType::String(s) => write!(f, "\"{}\"", s),
            TokenType::LeftShift => write!(f, "<<"),
            TokenType::RightShift => write!(f, ">>"),
            TokenType::Pipe => write!(f, "|"),
            TokenType::Carrot => write!(f, "^"),
            TokenType::Ampersand => write!(f, "&"),
            TokenType::KeywordStr => write!(f, "str"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub type_: TokenType,
    pub pos: (usize, usize),
}

impl Token {
    pub fn new(type_: TokenType, pos: (usize, usize)) -> Self {
        Self { type_, pos }
    }
}
