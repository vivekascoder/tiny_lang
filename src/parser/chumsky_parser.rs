use chumsky::prelude::*;

// fn parser<'a, I>() -> impl Parser<'a, I, SExpr, extra::Err<Rich<'a, Token<'a>>>>
// where
//     I: ValueInput<'a, Token = Token<'a>, Span = SimpleSpan>,
// {
//     recursive(|sexpr| {
//         // let atom = select! {
//         //     Token::Float(x) => SExpr::Float(x.parse().unwrap()),
//         //     Token::Add => SExpr::Add,
//         //     Token::Sub => SExpr::Sub,
//         //     Token::Mul => SExpr::Mul,
//         //     Token::Div => SExpr::Div,
//         // };

//         // let list = sexpr
//         //     .repeated()
//         //     .collect()
//         //     .map(SExpr::List)
//         //     .delimited_by(just(Token::LParen), just(Token::RParen));

//         // atom.or(list)
//     })
// }
