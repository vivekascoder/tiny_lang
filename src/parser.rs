use crate::ast::*;
use crate::lexer::Lexer;
use anyhow::{anyhow, bail, Result};
use log::info;
use std::rc::Rc;

pub struct Parser {
    lexer: Lexer,
    current_token: Rc<TokenType>,
    next_token: Rc<TokenType>,
    errors: Vec<anyhow::Error>,
}

impl Parser {
    pub fn new(module: &str, source: &str) -> Self {
        let mut parser = Self {
            lexer: Lexer::new(module, source),
            current_token: Rc::new(TokenType::EOF),
            next_token: Rc::new(TokenType::EOF),
            errors: vec![],
        };
        parser.bump().unwrap();
        parser.bump().unwrap();
        parser
    }

    pub fn module(&self) -> String {
        self.lexer.module()
    }

    pub fn get_row(&self) -> usize {
        self.lexer.get_row()
    }

    pub fn get_col(&self) -> usize {
        self.lexer.get_col()
    }

    fn bump(&mut self) -> Result<()> {
        self.current_token = Rc::clone(&self.next_token);
        self.next_token = Rc::new(self.lexer.next()?.type_);
        Ok(())
    }

    fn expect_next_token(&mut self, token: &Rc<TokenType>) -> Result<bool> {
        if self.next_token_is(token) {
            self.bump()?;
            Ok(true)
        } else {
            self.error_next_token(token);
            Ok(false)
        }
    }

    fn error_next_token(&mut self, tok: &TokenType) {
        self.errors.push(anyhow!(
            "expected {:?}, but got {:?}",
            tok,
            &self.next_token
        ))
    }

    fn next_token_is(&self, token: &Rc<TokenType>) -> bool {
        &self.next_token == token
    }
    fn current_token_is(&self, token: &Rc<TokenType>) -> bool {
        &self.current_token == token
    }

    fn parse_let_statement(&mut self) -> Result<Statement> {
        // the current token is `let`
        let name = match self.next_token.as_ref() {
            TokenType::Identifier(name) => name.clone(),
            _ => panic!("identifier isn't here."),
        };
        self.bump()?;

        // next token should be assign.
        if !self.expect_next_token(&Rc::new(TokenType::Equal))? {
            panic!("next token is not assign/equal.")
        }

        // next token should be expression.
        self.bump()?;

        // current token should be expression
        let expr = self.parse_expression(Precedence::Lowest)?;

        if !self.expect_next_token(&Rc::new(TokenType::SemiColon))? {
            bail!(
                "Line: {}, Col: {}, `;` not found",
                self.lexer.get_row(),
                self.lexer.get_col()
            )
        }
        self.bump()?;

        Ok(Statement::Let(Ident(Rc::clone(&name)), expr))
    }

    fn parse_expression_statement(&mut self) -> Result<Statement> {
        match self.parse_expression(Precedence::Lowest) {
            Ok(v) => {
                if !self.expect_next_token(&Rc::new(TokenType::SemiColon))? {
                    bail!("`;` not found at the end in expression statment.");
                }
                self.bump()?;
                Ok(Statement::Expr(v))
            }
            Err(e) => {
                bail!("error while parsing expression statement: {:?}", e);
            }
        }
    }

    fn keyword_to_type(&self, tok: &TokenType) -> Result<Type> {
        match tok {
            TokenType::KeywordUsize => Ok(Type::UnsignedInteger),
            TokenType::KeywordBool => Ok(Type::Bool),
            _ => {
                bail!("{:?} is not a valid parameter type.", tok);
            }
        }
    }

    fn parse_function_statement(&mut self) -> Result<Statement> {
        // current token is `fun`.

        let fn_name = match self.next_token.as_ref() {
            TokenType::Identifier(n) => Rc::clone(&n),
            _ => {
                bail!(
                    "next token to parse function should be Identifier, got {:?} instead",
                    &self.next_token
                );
            }
        };
        self.bump()?;

        // TODO: use `expect_next_token()`
        if !self.next_token_is(&Rc::new(TokenType::LParen)) {
            bail!(
                "next token to parse function should be LParen, got {:?} instead",
                &self.next_token
            );
        }
        self.bump()?;

        let mut params: Vec<(Ident, Type)> = vec![];

        // Start parsing parameters
        while !self.current_token_is(&Rc::new(TokenType::RParen)) {
            // parse ident-> : ->type
            let param_name = match self.next_token.as_ref() {
                TokenType::Identifier(ref n) => Rc::clone(n),
                _ => {
                    bail!("next token to parse function parameters should be an Identifier, got {:?} instead", &self.next_token);
                }
            };
            self.bump()?;

            if !self.next_token_is(&Rc::new(TokenType::Colon)) {
                bail!(
                    "next token to parse function parameters should be an Colon, got {:?} instead",
                    &self.next_token
                );
            }
            self.bump()?;

            let type_ = self.keyword_to_type(&self.next_token)?;
            self.bump()?;

            if !(self.next_token_is(&Rc::new(TokenType::Comma))
                || self.next_token_is(&Rc::new(TokenType::RParen)))
            {
                bail!("expected Command, but found {:?}", self.next_token);
            }
            self.bump()?;

            params.push((Ident(param_name), type_));
        }

        // parse the optional return type.
        if !self.expect_next_token(&Rc::new(TokenType::SymbolReturn))? {
            bail!("Expected `=>` but got {:?}", self.next_token);
        }

        let mut return_type: Option<Type> = None;
        info!("Next token is: {:?}", self.next_token);
        if !self.next_token_is(&Rc::new(TokenType::KeywordVoid)) {
            return_type = Some(self.keyword_to_type(&self.next_token)?);
        }
        self.bump()?;

        if !self.expect_next_token(&Rc::new(TokenType::LBrace))? {
            bail!("next token is not `{{`");
        }

        self.bump()?;

        let body = self.parse_block_statement()?;

        Ok(Statement::Function(Function {
            name: fn_name,
            params: params,
            return_type: return_type,
            body: body,
        }))
    }

    fn parse_block_statement(&mut self) -> Result<BlockStatement> {
        let mut statements: BlockStatement = vec![];

        info!("Current token: {:?}", &self.current_token);

        while !self.current_token_is(&Rc::new(TokenType::RBrace)) {
            statements.push(self.parse_statement()?);
        }

        self.bump()?;

        Ok(statements)
    }

    fn parse_return_statement(&mut self) -> Result<Statement> {
        self.bump()?;
        let expr = self.parse_expression(Precedence::Lowest)?;

        if !self.expect_next_token(&Rc::new(TokenType::SemiColon))? {
            bail!("return statement has no semicolon");
        }
        self.bump()?;

        Ok(Statement::Return(expr))
    }

    fn parse_if_statement(&mut self) -> Result<Statement> {
        if !self.expect_next_token(&Rc::new(TokenType::LParen))? {
            bail!(
                "parsing if: expected `(` but found {:?}, instead",
                self.current_token
            );
        }

        // WHY: to make the current token not `(`
        self.bump()?;

        // condition
        let expr = self.parse_expression(Precedence::Lowest)?;
        info!("Expression: {:?}", expr);

        if !self.expect_next_token(&Rc::new(TokenType::RParen))? {
            bail!(
                "parsing if: expected `)` but found {:?}, instead",
                self.current_token
            );
        }

        if !self.expect_next_token(&Rc::new(TokenType::LBrace))? {
            bail!(
                "parsing if: expected `{{` but found {:?}, instead",
                self.current_token
            );
        }
        self.bump()?;

        let block = self.parse_block_statement()?;

        // If next token is `else` then parse else block.
        let mut else_body: Option<BlockStatement> = None;
        if self.current_token_is(&Rc::new(TokenType::KeywordElse)) {
            if !self.expect_next_token(&Rc::new(TokenType::LBrace))? {
                bail!(
                    "expected `{{` while parsing else body but got {:?}",
                    &self.next_token
                );
            }
            self.bump()?;
            else_body = Some(self.parse_block_statement()?);
        }

        Ok(Statement::If(Condition {
            condition: expr,
            if_body: block,
            else_body: else_body,
        }))
    }

    fn parse_while_statement(&mut self) -> Result<Statement> {
        if !self.expect_next_token(&Rc::new(TokenType::LParen))? {
            bail!("expected `(` but got {:>} instead", self.next_token);
        }
        self.bump()?;

        let expr = self.parse_expression(Precedence::Lowest)?;
        info!("Expression: {:?}", expr);

        if !self.expect_next_token(&Rc::new(TokenType::RParen))? {
            bail!(
                "parsing if: expected `)` but found {:?}, instead",
                self.current_token
            );
        }

        if !self.expect_next_token(&Rc::new(TokenType::LBrace))? {
            bail!(
                "parsing if: expected `{{` but found {:?}, instead",
                self.current_token
            );
        }
        self.bump()?;

        let block = self.parse_block_statement()?;

        info!("while statment parsed: {:?}, {:?}", &expr, &block);

        Ok(Statement::While(While {
            condition: expr,
            body: block,
        }))
    }

    fn parse_assign_or_expr(&mut self) -> Result<Statement> {
        let var = match self.current_token.as_ref() {
            TokenType::Identifier(ref i) => Rc::clone(i),
            _ => {
                bail!("{:?} is not identifier", self.current_token);
            }
        };

        if self.next_token_is(&Rc::new(TokenType::Equal)) {
            // parse assignment.
            self.bump()?;
            self.bump()?;
            let expr = self.parse_expression(Precedence::Lowest)?;

            info!("=/i {:?}, {:?}", self.current_token, self.next_token);

            if !self.expect_next_token(&Rc::new(TokenType::SemiColon))? {
                bail!("end plx");
            }
            self.bump()?;
            Ok(Statement::Assignment(Ident(var), expr))
        } else {
            Ok(self.parse_expression_statement()?)
        }
    }

    fn parse_statement(&mut self) -> Result<Statement> {
        info!("Current Token: {:?}", self.current_token);
        Ok(match self.current_token.as_ref() {
            TokenType::KeywordLet => self.parse_let_statement()?,
            TokenType::KeywordIf => self.parse_if_statement()?,
            TokenType::KeywordFun => self.parse_function_statement()?,
            TokenType::KeywordWhile => self.parse_while_statement()?,
            TokenType::KeywordReturn => self.parse_return_statement()?,
            TokenType::Identifier(_) => self.parse_assign_or_expr()?,
            _ => self.parse_expression_statement()?,
        })
    }

    fn token_to_precedence(tok: &TokenType) -> Precedence {
        match tok {
            TokenType::Equal | TokenType::NotEqual | TokenType::DoubleEqual => Precedence::Equals,
            TokenType::LessThan | TokenType::LessThanEqual => Precedence::LessGreater,
            TokenType::GreaterThan | TokenType::GreaterThanEqual => Precedence::LessGreater,
            TokenType::Plus | TokenType::Minus => Precedence::Sum,
            TokenType::Divide | TokenType::Multiply => Precedence::Product,
            TokenType::LBrace => Precedence::Index,
            TokenType::LParen => Precedence::Call,
            _ => Precedence::Lowest,
        }
    }

    fn current_token_precedence(&self) -> Precedence {
        Self::token_to_precedence(&self.current_token)
    }

    fn next_token_precedence(&self) -> Precedence {
        Self::token_to_precedence(&self.next_token)
    }

    fn parse_expression(&mut self, precedence: Precedence) -> Result<Expr> {
        // Parse the left
        let mut left = match self.current_token.as_ref() {
            TokenType::Usize(val) => Expr::Literal(Literal::UnsignedInteger(*val)),
            TokenType::Boolean(val) => Expr::Literal(Literal::Bool(*val)),
            TokenType::Identifier(ref i) => Expr::Ident(Ident(i.clone())),
            TokenType::SQuote => {
                use TokenType::*;
                // Parse character.
                let char_val = match self.next_token.as_ref() {
                    Identifier(ref i) => {
                        if i.len() != 1 {
                            bail!("Character should be of length 1");
                        }
                        Rc::clone(i)
                    }

                    // TODO: could use display here.
                    NewLine => Rc::from('\n'.to_string()),
                    Tab => Rc::from('\t'.to_string()),
                    Space => Rc::from(' '.to_string()),
                    BSlash => Rc::from('\\'.to_string()),
                    Minus => Rc::from('-'.to_string()),

                    _ => {
                        bail!("char not found after `'` instead got {:?}", self.next_token);
                    }
                };
                self.bump()?;
                if !self.expect_next_token(&Rc::new(TokenType::SQuote))? {
                    bail!(
                        "character not ended with `'` found {:?} instead.",
                        self.next_token
                    );
                }
                Expr::Literal(Literal::Char(char_val.chars().nth(0).unwrap()))
            }
            TokenType::Minus | TokenType::Bang | TokenType::Plus => {
                // Parse prefix expression.

                // Operator
                let prefix = match self.current_token.as_ref() {
                    TokenType::Bang => Prefix::Not,
                    TokenType::Plus => Prefix::Plus,
                    TokenType::Minus => Prefix::Minus,
                    _ => bail!("not a valid infix operator."),
                };
                self.bump()?;

                // Expression
                match self.parse_expression(Precedence::Prefix) {
                    Ok(val) => {
                        return Ok(Expr::Prefix(prefix, Box::new(val)));
                    }
                    Err(e) => bail!("error while parsing expression for prefix with {:#?}", e),
                }
            }

            _ => {
                bail!(
                    "no prefix parse function found for {:?}",
                    self.current_token
                );
            }
        };
        info!("Left: {:?}", left);

        // Parse the infix
        while !self.next_token_is(&Rc::new(TokenType::SemiColon))
            && precedence < self.next_token_precedence()
        {
            info!(
                "next token while parsing expression right -> {:?}",
                self.next_token
            );
            match self.next_token.as_ref() {
                TokenType::Plus
                | TokenType::Minus
                | TokenType::Divide
                | TokenType::Multiply
                | TokenType::Equal
                | TokenType::DoubleEqual
                | TokenType::NotEqual
                | TokenType::LessThan
                | TokenType::LessThanEqual
                | TokenType::GreaterThan
                | TokenType::GreaterThanEqual => {
                    self.bump()?;

                    let infix = match self.current_token.as_ref() {
                        TokenType::Plus => Infix::Plus,
                        TokenType::Minus => Infix::Minus,
                        TokenType::Divide => Infix::Divide,
                        TokenType::Multiply => Infix::Multiply,
                        TokenType::DoubleEqual => Infix::DoubleEqual,
                        TokenType::NotEqual => Infix::NotEqual,
                        TokenType::LessThan => Infix::LessThan,
                        TokenType::LessThanEqual => Infix::LessThanEqual,
                        TokenType::GreaterThan => Infix::GreaterThan,
                        TokenType::GreaterThanEqual => Infix::GreaterThanEqual,
                        _ => {
                            bail!("{:?} is not a valid infix operator.", &self.next_token);
                        }
                    };
                    info!("Infix: {:?}", infix);

                    let precedence = self.current_token_precedence();
                    self.bump()?;

                    left = match self.parse_expression(precedence) {
                        Ok(expr) => Expr::Infix(infix, Box::new(left), Box::new(expr)),
                        Err(e) => {
                            bail!("error while parsing expr: {:?}", e);
                        }
                    }
                }
                TokenType::LParen => {
                    let mut params: Vec<Expr> = vec![];
                    self.bump()?;

                    while !self.current_token_is(&Rc::new(TokenType::RParen)) {
                        self.bump()?;
                        let param = self.parse_expression(Precedence::Lowest)?;
                        info!("param: {:?}", param);
                        params.push(param);

                        info!("{:?}, {:?}", self.current_token, self.next_token);
                        if !(self.next_token_is(&Rc::new(TokenType::Comma))
                            || self.next_token_is(&Rc::new(TokenType::RParen)))
                        {
                            bail!("sep , while param func");
                        }
                        self.bump()?;
                        info!(" -->{:?}, {:?}", self.current_token, self.next_token);
                    }

                    // self.bump()?;
                    info!(" -->{:?}, {:?}", self.current_token, self.next_token);
                    // self.bump()?;

                    if let Expr::Ident(i) = left {
                        left = Expr::Call(FunctionCall {
                            parameters: params,
                            name: i.0,
                        })
                    } else {
                        bail!("left is not an identifier.");
                    }
                }
                _ => return Ok(left),
            }
        }

        Ok(left)
    }

    /// main function to parse a lexer.
    pub fn parse(&mut self) -> Result<Program> {
        let mut program: Program = vec![];
        while !self.current_token_is(&Rc::new(TokenType::EOF)) {
            match self.parse_statement() {
                Ok(stmt) => {
                    info!("Statement: {:?}", &stmt);
                    program.push(stmt);
                }
                Err(e) => {
                    bail!("That's it: {:?}", e)
                }
            }
        }
        Ok(program)
    }
}
