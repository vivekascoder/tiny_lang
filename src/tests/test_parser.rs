#[cfg(test)]
pub mod tests {
    use crate::ast::*;
    use crate::parser::Parser;

    #[test]
    fn test_let_statement_parsing() {
        let code = "let something = 454 + 3 * 4 - 35;";
        let mut parser = Parser::new(code);
        // println!("Parsed statements: {:#?}", parser.parse());
        assert_eq!(
            parser.parse().unwrap(),
            [Statement::Let(
                Ident("something".to_string()),
                Expr::Infix(
                    Infix::Minus,
                    Box::new(Expr::Infix(
                        Infix::Plus,
                        Box::new(Expr::Literal(Literal::UnsignedInteger(454,))),
                        Box::new(Expr::Infix(
                            Infix::Multiply,
                            Box::new(Expr::Literal(Literal::UnsignedInteger(3))),
                            Box::new(Expr::Literal(Literal::UnsignedInteger(4))),
                        )),
                    )),
                    Box::new(Expr::Literal(Literal::UnsignedInteger(35,)))
                ),
            )]
        );
    }

    #[test]
    fn test_prefix_expression_parsing() {
        let code = "let val = +454;";
        let mut parser = Parser::new(code);
        // println!("Parsed statement: {:#?}", parser.parse());
        assert_eq!(
            parser.parse().unwrap(),
            [Statement::Let(
                Ident("val".to_string()),
                Expr::Prefix(
                    Prefix::Plus,
                    Box::new(Expr::Literal(Literal::UnsignedInteger(454)))
                )
            )]
        )
    }

    fn assert_eq_ast(code: &str, ast: Program) {
        assert_eq!(Parser::new(code).parse().unwrap(), ast);
    }

    #[test]
    fn test_function_parsing() {
        let code = r#"
        fun return_something(a: usize) => bool {
            let something = false;
            let another_var = 3535 + 35;
            return something;
        }
        "#;
        let mut parser = Parser::new(code);

        // println!("Parsed: {:#?}", parser.parse());
        assert_eq!(
            parser.parse().unwrap(),
            [Statement::Function(Function {
                name: "return_something".to_string(),
                params: vec![(Ident("a".to_string()), Type::UnsignedInteger)],
                return_type: Some(Type::Bool),
                body: vec![
                    Statement::Let(
                        Ident("something".to_string()),
                        Expr::Literal(Literal::Bool(false))
                    ),
                    Statement::Let(
                        Ident("another_var".to_string()),
                        Expr::Infix(
                            Infix::Plus,
                            Box::new(Expr::Literal(Literal::UnsignedInteger(3535))),
                            Box::new(Expr::Literal(Literal::UnsignedInteger(35)))
                        )
                    ),
                    Statement::Return(Expr::Ident(Ident("something".to_string()))),
                ],
            })]
        )
    }

    #[test]
    fn test_if_statement_parsing() {
        let code = r#"
        if (54 > 34) {
            let something = 446;
        } else {
            let somethine = 0;
        }
        "#;
        let ast: Program = vec![Statement::If(Condition {
            condition: Expr::Infix(
                Infix::GreaterThan,
                Box::new(Expr::Literal(Literal::UnsignedInteger(54))),
                Box::new(Expr::Literal(Literal::UnsignedInteger(34))),
            ),
            if_body: vec![Statement::Let(
                Ident("something".to_string()),
                Expr::Literal(Literal::UnsignedInteger(446)),
            )],
            else_body: Some(vec![Statement::Let(
                Ident("somethine".to_string()),
                Expr::Literal(Literal::UnsignedInteger(0)),
            )]),
        })];
        assert_eq_ast(code, ast);
    }
}
