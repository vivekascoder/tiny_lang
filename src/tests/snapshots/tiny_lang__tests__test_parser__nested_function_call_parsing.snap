---
source: src/tests/test_parser.rs
expression: "Parser::new(code).parse()"
---
Ok(
    [
        Expr(
            Call(
                FunctionCall {
                    parameters: [
                        Infix(
                            Plus,
                            Literal(
                                UnsignedInteger(
                                    35,
                                ),
                            ),
                            Literal(
                                UnsignedInteger(
                                    3435,
                                ),
                            ),
                        ),
                        Call(
                            FunctionCall {
                                parameters: [
                                    Literal(
                                        UnsignedInteger(
                                            34535,
                                        ),
                                    ),
                                    Literal(
                                        UnsignedInteger(
                                            355,
                                        ),
                                    ),
                                ],
                                name: "another_func",
                            },
                        ),
                    ],
                    name: "function_call",
                },
            ),
        ),
    ],
)
